### DISCUSSION TITLE
How do I access `sci.core/eval-string*` inside of babushka?

### DISCUSSION THREAD

**Discussion by kkinnear** [babashka/babashka#1590] [Q&A] [answered]
In my library zprint, I use `sci`  (specifically `sci.core/eval-string*`) to turn maps (which sometimes contain functions) read in from files into Clojure maps.  `sci` works great, and produces functions which work fine but are also essentially "sandboxed", and so won't do terrible things to someone's environment.  

When I try to run the zprint library inside of babushka, I can't seem to figure out how to access `sci.core/eval-string*` in order to do the same thing.  If I just use `read-string`, it will read a string but not produce a working function when one appears in the string.  I'm using the reader-conditional `:bb` to do all of this, for what that's worth.

I know that babushka is implemented with `sci`, so it isn't like `sci` isn't "in there" somewhere, but despite plenty of trial and error, I can't seem to figure out any way to get the equivalent functionality to my calls to `sci.core/eval-string*` when actually running in babushka as opposed to Clojure or Clojurescipt.  To be clear, since I'm using reader-conditionals, I can do whatever is necessary in babushka without impacting the normal usage in the Clojure and Clojurescript environments.  So any equivalent to `sci.core/eval-string*` would be fine.  The only other `sci` thing I even use is `sci.core/init` to make some of my functions available to `sci` processed input.

Any help would be greatly appreciated!

**Comment by borkdude** [comment]
`load-string` in babashka will do the same thing as `sci.core/eval-string*`

**Reply by borkdude** [reply]
Another way to write bb-specific code compared to JVM is:

``` clojure
(defmacro if-bb [then else]
  (if (System/getProperty "babashka.version") then else))

(if-bb (def x 10) (def x 11))

x ;; => 10 in bb, 11 in JVM
```

**Reply by kkinnear** [reply]
Thank you so much!  `load-string` work great -- except for one detail.  When I use `sci.core/eval-string*`, I'm giving it a `ctx` that I created with `sci.core/init`.  Is there some way to make `load-string` aware of my `ctx`?  Or the information that I would like to put in a `ctx`, more specifically.

I've looked at the code in `sci` for `load-string`, and while the `impl` version seems to take a `ctx`, it looks like the `load-string` that is exposed has one given to it.  Not that I can call `init` to create a `ctx` anyway in babashka. Perhaps there is some way to get the information that I would like to have in a `ctx` known by babashka or sci prior to the call to `load-string`?  

Near as I can tell, this is the only thing missing in order to have a fully functional zprint that will work inside of babashka.  Which would be a neat thing in general, and particularly useful for the testing environment that I'm working for for zprint itself (which is the source of my current questions).

Thanks!

**Reply by borkdude** [reply]
`load-string` in babashka has an implicit context, the global context which is babashka's context. but maybe I could just expose SCI's interface (or parts of it) to babashka. Which functions are you using? `init`, `eval-string*`, more?

**Reply by kkinnear** [reply]
Wow, that would be great.  I use only `init` and `eval-string*`.   It would be really nice to be able to access these functions in babashka!

Thank you so much for `sci`! `sci` has been a wonderful way to handle processing configuration files for zprint.  It just works (which is no small thing), and protects people from bad things that might show up in the files.  It was trivial to integrate into zprint, and is the only library I use other than `rewrite-clj`.  It is an amazing piece of work in it own right, to say nothing of enabling babashka as well. Thank you for all of your excellent work on both of these Clojure capabilities that are becoming fundamental!

**Reply by borkdude** [reply, marked as answer]
I've exposed sci.core on master. You can test this with:

```
bash <(curl https://raw.githubusercontent.com/babashka/babashka/master/install) --dev-build --dir /tmp
```

as soon as the master build finishes.

**Reply by kkinnear** [reply]
Thank you very very much!

That's a very impressive string to download the snapshot version to test.  The download worked flawlessly!

```
√ projects/zprint % /tmp/bb --version
babashka v1.3.183-SNAPSHOT
√ projects/zprint %
```

And indeed, the zprint library also now works as well! I can now use the same code for babashka as for Clojure (no reader-conditionals necessary), and when I read in a string with `sci.core/eval-string*` it finds functions that I have added to a context using `sci.core/init`.  That's great, and I will be letting people know that zprint now works completely when used within babashka.

One final question -- is there some way to add to the internal `sci` context within babashka itself?  I can certainly put everything in the `ns` statement at the top of the babashka file, but if I have 15 functions or so, and they are all set up to be fed to `sci.core/init`, is there something I can call at the start of my file to add them to the internal babashka `sci` context?  This isn't critical, because I can certainly make babashka aware of them using the `ns` statement, but if there is some way to add them to the internal babashka `sci` context, that would be convenient.

Thank you again for making all of this possible.  It is a huge help to me in creating a system test for zprint using babashka, and I think it will be a useful thing for others that might want to use zprint inside of babashka.  To say nothing of other folks or libraries that use `sci` that can now have them operate within babashka unchanged!

**Reply by borkdude** [reply]
The internal SCI context of babashka isn't exposed so I recommend doing it via the ns form.

**Reply by kkinnear** [reply]
Great thanks.  Thanks again for exposing `sci.core`!
